include/dgl/immutable_graph.h

```diff
@@ -930,100 +917,39 @@ class ImmutableGraph: public GraphInterface {
-      return ImmutableGraph(nullptr, csr);
-    } else {
-      LOG(FATAL) << "Unknown edge direction: " << edge_dir;
-      return ImmutableGraph();
-    }
-  }
-
-  static ImmutableGraph CreateFromCSR(IdArray indptr, IdArray indices, IdArray edge_ids,
-                                      bool multigraph, const std::string &edge_dir) {
-    CSRPtr csr(new CSR(indptr, indices, edge_ids, multigraph));
-    if (edge_dir == "in") {
-      return ImmutableGraph(csr, nullptr);
-    } else if (edge_dir == "out") {
-      return ImmutableGraph(nullptr, csr);
-    } else {
-      LOG(FATAL) << "Unknown edge direction: " << edge_dir;
-      return ImmutableGraph();
-    }
-  }
-
-  static ImmutableGraph CreateFromCSR(IdArray indptr, IdArray indices, IdArray edge_ids,
-                                      const std::string &edge_dir,
-                                      const std::string &shared_mem_name) {
-    CSRPtr csr(new CSR(indptr, indices, edge_ids, GetSharedMemName(shared_mem_name, edge_dir)));
-    if (edge_dir == "in") {
-      return ImmutableGraph(csr, nullptr, shared_mem_name);
-    } else if (edge_dir == "out") {
-      return ImmutableGraph(nullptr, csr, shared_mem_name);
-    } else {
-      LOG(FATAL) << "Unknown edge direction: " << edge_dir;
-      return ImmutableGraph();
-    }
-  }
+  static ImGraphPtr AsNumBits(ImGraphPtr g, uint8_t bits);
 
-  static ImmutableGraph CreateFromCSR(IdArray indptr, IdArray indices, IdArray edge_ids,
-                                      bool multigraph, const std::string &edge_dir,
-                                      const std::string &shared_mem_name) {
-    CSRPtr csr(new CSR(indptr, indices, edge_ids, multigraph,
-                       GetSharedMemName(shared_mem_name, edge_dir)));
-    if (edge_dir == "in") {
-      return ImmutableGraph(csr, nullptr, shared_mem_name);
-    } else if (edge_dir == "out") {
-      return ImmutableGraph(nullptr, csr, shared_mem_name);
-    } else {
-      LOG(FATAL) << "Unknown edge direction: " << edge_dir;
-      return ImmutableGraph();
-    }
-  }
-
-  static ImmutableGraph CreateFromCSR(const std::string &shared_mem_name, size_t num_vertices,
-                                      size_t num_edges, bool multigraph,
-                                      const std::string &edge_dir) {
-    CSRPtr csr(new CSR(GetSharedMemName(shared_mem_name, edge_dir), num_vertices, num_edges,
-                       multigraph));
-    if (edge_dir == "in") {
-      return ImmutableGraph(csr, nullptr, shared_mem_name);
-    } else if (edge_dir == "out") {
-      return ImmutableGraph(nullptr, csr, shared_mem_name);
-    } else {
-      LOG(FATAL) << "Unknown edge direction: " << edge_dir;
-      return ImmutableGraph();
-    }
-  }
+  /*!
+   * \brief Return a new graph with all the edges reversed.
+   *
+   * The returned graph preserves the vertex and edge index in the original graph.
+   *
+   * \return the reversed graph
+   */
+  ImGraphPtr Reverse() const;
```

Comment:
`func() const` are functions that don't change internal state of the object.
`static` are just for class methods.  I'm also not sure what is the general guideline of member v.s. static v.s. global functions.  [Google has a C++ guideline](https://google.github.io/styleguide/cppguide.html#Nonmember,_Static_Member,_and_Global_Functions) but we don't seem to follow it anyway.

Reply:
We should try follow the Google coding style. I agree that in general we should avoid grouping a bunch of static functions such as in `GraphOp`. The reason it happened (and perhaps we should fix it) is that some operations are more efficient if they can directly access the internal state of a graph object. Again, this is bad practice as it breaks the encapsulation. We should get rid of the `GraphOp` bundle class. `Reverse` here is similar because it needs to know whether CSR or COO has been materialized to choose a more efficient implementation. I think a better way to handle this is to add methods to indicate whether CSR or COO has been materialized so we could move the logic outside and make `Reverse` a global function. Let's do that gradually in the following PRs.